subprojects/file-watching/src/integTest/groovy/org/gradle/internal/watch/ChangesBetweenBuildsFileSystemWatchingIntegrationTest.groovy

```diff
@@ -169,4 +171,70 @@ class ChangesBetweenBuildsFileSystemWatchingIntegrationTest extends AbstractFile
         outputContains "Hello VFS!"
         executedAndNotSkipped ":compileJava", ":classes", ":run"
     }
+
+    @Issue("https://github.com/gradle/gradle/issues/17865")
+    def "project directory moved is handled properly"() {
+        def buildCache = new TestBuildCache(temporaryFolder.file("cache-dir").deleteDir().createDir())
+        def buildFileContents = """
+            apply plugin: "base"
+
+            @CacheableTask
+            abstract class CustomTask extends DefaultTask {
+                @InputFile
+                @PathSensitive(PathSensitivity.NONE)
+                abstract RegularFileProperty getSource()
+
+                @OutputFile
+                abstract RegularFileProperty getTarget()
+
+                @TaskAction
+                def execute() {
+                    target.get().asFile.text = source.get().asFile.text
+                }
+            }
+
+            tasks.register("run", CustomTask) {
+                source = file("source.txt")
+                target = file("build/target.txt")
+            }
+        """
+
+        def originalDir = file("original")
+        def renamedDir = file("renamed")
+        def settingsFile = originalDir.file("settings.gradle")
+        def buildFile = originalDir.file("build.gradle")
+        def sourceFile = originalDir.file("source.txt")
+        def targetFile = originalDir.file("build/target.txt")
+
+        executer.beforeExecute {
+            inDirectory(originalDir)
+            enableVerboseVfsLogs()
+            withBuildCache()
+            withWatchFs()
+        }
+
+        settingsFile.text = buildCache.localCacheConfiguration()
+        buildFile.text = buildFileContents
+        sourceFile.text = "Hello Old World!"
+
+        when:
+        run "run"
+        then:
+        targetFile.text == "Hello Old World!"
+        executedAndNotSkipped ":run"
+
+        expect:
+        succeeds "clean"
+
+        when:
+        originalDir.renameTo(renamedDir)
```

Comment:
Does this mean that the issue isn't actually Windows related?

Reply:
what if we do a case-only rename in Windows?

Reply:
The issue is Windows-related, but there is no point in keeping the test coverage Windows-only.

We will handle a case-rename the same way as any other rename, and report it as moved. This is because `GetFinalPathNameByHandleW()` will return a different path after the rename.